experiment/compatibility-versions/validate-compatibility-versions-feature-gates.sh

```diff
@@ -0,0 +1,245 @@
+#!/bin/bash
+# Copyright 2025 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script validates feature gates from a Kubernetes cluster's /metrics endpoint
+# against expected values defined in versioned and unversioned feature lists.
+#
+# Usage: validate_feature_gates.sh <emulated_version> <metrics_file> <versioned_feature_list> <unversioned_feature_list> <results_file>
+
+set -o errexit -o nounset -o pipefail
+
+validate_feature_gates() {
+  local emulated_version="$1"   # e.g. "1.32"
+  local metrics_file="$2"       # path to /metrics
+  local feature_list="$3"       # versioned_feature_list.yaml
+  local unversioned_feature_list="$4" # unversioned_feature_list.yaml
+  local results_file="$5"
+
+  echo "Validating features for ${emulated_version}..."
+  rm -f "${results_file}"
+  touch "${results_file}"
+
+  # Parse /metrics -> actual_features[featureName] = 0 or 1
+  declare -A actual_features
```

Comment:
worth noting: when we have a lot of arrays that aren't something trivial like exec args that's a sign that we probably want to reach for a more structured language

Reply:
+1, I originally wrote this test in Golang but ported it to bash for use in test-infra:
https://github.com/kubernetes/kubernetes/pull/128138

Reply:
ACK, I commented there to continue the conversation 👍